### PR TITLE
Invalidate script module cache when the underlying module file is changed

### LINKED ISSUE #2505: Reloading module does not reload submodules.
## Steps to reproduce

``` powershell
#Sub.psm1:
  class Sub {
    [int] $Value
  }

#modrepo.psm1:
  using module .\Sub.psm1

  function Get-Sub { 
     [Sub]@{
         Value = 42
     }
  }

Import-Module modrepo
Get-Sub
```

`Value : 1`

Change the definition of Sub:

``` powershell
#Sub.psm1:
  class Sub {  
    [string] $Name # added field
    [int] $Value
  }

#modrepo.psm1:
  using module .\Sub.psm1

  function Get-Sub{ 
     [Sub]@{
         Name = ‘Staffan’ # added field
         Value = 42
     }
  }

#Then

Remove-Module -force modrepo
Import-Module -force modrepo
Get-Sub 
```
## Expected behavior

output of object of class Sub with 

```
Name : 'Staffan'
Value : 42
```
## Actual behavior

```
Cannot create object of type "Sub". The Name property was not found for the Sub object. The available property is: [Value <System.Int32>]
At C:\Users\<user>\Documents\WindowsPowerShell\Modules\modrepo\modrepo.psm1:9 char:3
+   [Sub] @{
+   ~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [], RuntimeException
    + FullyQualifiedErrorId : ObjectCreationError
```
## Environment data

``` powershell

Name                           Value                                                                                                                                                       
----                           -----                                                                                                                                                       
PSVersion                      5.1.14393.206                                                                                                                                               
PSEdition                      Desktop                                                                                                                                                     
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                     
BuildVersion                   10.0.14393.206                                                                                                                                              
CLRVersion                     4.0.30319.42000                                                                                                                                             
WSManStackVersion              3.0                                                                                                                                                         
PSRemotingProtocolVersion      2.3                                                                                                                                                         
SerializationVersion           1.1.0.1                                                                                                                                                     



```

### PR DESCRIPTION
**Pull request by rjmholt**
## PR Summary

Fix the cache invalidation issue from #2505.

Currently, importing a module for the second time will always import from the cache, even if the underlying file has been changed.

This change makes the cache check the last write time on the underlying file first before proceeding, and if it is different, updates the cache and returns the new module file.

I have also changed the analysis cache dictionaries to use `ConcurrentDictionary` rather than needing to manually use a locking object.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/powershell/6710)

### CODE REVIEW

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 1]
```diff
@@ -5809,10 +5805,7 @@ private static BinaryAnalysisResult GetCmdletsFromBinaryModuleImplementation(str
         {
             Tuple<BinaryAnalysisResult, Version> tuple;
 
-            lock (s_lockObject)
-            {
-                s_binaryAnalysisCache.TryGetValue(path, out tuple);
-            }
+            s_binaryAnalysisCache.TryGetValue(path, out tuple);
 
             if (tuple != null)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: Seems we could do
```c#
if (s_binaryAnalysisCache.TryGetValue(path, out tuple))
```

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 2]
```diff
@@ -5845,10 +5835,7 @@ private BinaryAnalysisResult GetCmdletsFromBinaryModuleImplementation(string pat
         {
             Tuple<BinaryAnalysisResult, Version> tuple = null;
 
-            lock (s_lockObject)
-            {
-                s_binaryAnalysisCache.TryGetValue(path, out tuple);
-            }
+            s_binaryAnalysisCache.TryGetValue(path, out tuple);
 
             if (tuple != null)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: The same.

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 3]
```diff
@@ -5986,29 +5970,42 @@ private static void AnalyzeSnapinDomainHelper()
...
+            ScriptModuleCacheEntry cacheEntry = null;
+            s_scriptAnalysisCache.TryGetValue(filePath, out cacheEntry);
 
-            lock (s_lockObject)
+            // Now check the file to see if its changed since we cached it
+            DateTime lastModuleWriteTimeUtc;
+            TryGetFileLastWriteTimeUtc(filePath, out lastModuleWriteTimeUtc);
+
+            // Return here if the cache is up to date
+            if (cacheEntry?.ModuleInfo != null && lastModuleWriteTimeUtc == cacheEntry.LastFileWriteTimeUtc)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: About `TryGetFileLastWriteTimeUtc` see my comment below. We could do:
```c#
var lastModuleWriteTimeUtc = GetFileLastWriteTimeUtc(filePath)
if (s_scriptAnalysisCache.TryGetValue(filePath, out ScriptModuleCacheEntry cacheEntry) && cacheEntry.ModuleInfo != null && lastModuleWriteTimeUtc == cacheEntry.LastFileWriteTimeUtc)
{

}
```

And I wonder why here `ModuleInfo` could be null?

File: src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs [hunk 4]
```diff
@@ -7252,6 +7243,58 @@ internal static PSSnapInInfo GetEngineSnapIn(ExecutionContext context, string na
...
+            try
+            {
+                lastWriteTimeUtc = File.GetLastAccessTimeUtc(path);
+                return true;
+            }
+            catch (Exception e)
+            {
+                ModuleIntrinsics.Tracer.WriteLine("Exception checking LastWriteTimeUtc on module {0}: {1}", path, e.Message);
+                lastWriteTimeUtc = DateTime.MinValue;
+                return false;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/Modules/ModuleCmdletBase.cs]: We return `lastWriteTimeUtc` in any case. So I suggest rename to `GetFileLastWriteTimeUtc`, don't return a bool and directly return `GetFileLastWriteTimeUtc`
```c#
internal static DateTime GetFileLastWriteTimeUtc(string path) 
```

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,518 @@
...
+    $modDir = New-Item -Path $modBasePath -ItemType Directory -Force
+    $null = New-Item -Path $modPath -Value $Content -Force
+
+    $null = $script:ModuleNames.Add($modName)
+
+    @{
+        "BaseDir" = $modDir;
+        "Name" = $modName;
+        "Path" = $modPath
+    }
```
Comment (iSazonov) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Could we use `return` in all helper functions?

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+            $module = Import-Module $modData.Path -PassThru -Force
+            Test-SecondModuleFunction | Should -BeExactly "SECONDSTRING"
+        }
+
+        It "Re-imports modules from file with new members when -Force is used with -PassThru" {
+            $modData = New-TestModule -Content $simpleModContent
+            $module = Import-Module $modData.Path -PassThru
+
+            $module.ExportedFunctions.Count                                    | Should -Be 1
+            $module.ExportedFunctions.ContainsKey("Test-FirstModuleFunction")  | Should -BeTrue
```
Comment (iSazonov) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Please use the same pattern from Line 128 with `Should -Contain`. Below too.

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,518 @@
...
+
+            $mod1 = Import-Module $modData.Path -PassThru
+
+            Set-Content -Path $modData.Path -Value $modSrc2
+            $mod2 = Import-Module $modData.Path -PassThru
+
+            $firstTypes = $mod1.GetExportedTypeDefinitions()
+            $secondTypes = $mod2.GetExportedTypeDefinitions()
+
+            $firstTypes.MyObj.Equals($secondTypes.Sub) | Should -BeFalse
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Where does the `Sub` come from?

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+            )
+
+            $simpleModContent = @'
+function Test-FirstModuleFunction
+{
+    Write-Output "TESTSTRING"
+}
+'@
+
+            $ps = [powershell]::Create()
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Why creating the powershell instance if it's never used?

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    It "Re-imports modules from file with new members when -Force is used" {
+        $module = Import-Module $modInfo -PassThru
+
+        $module.ExportedFunctions.Count | Should -Be 1
+        $module.ExportedFunctions.Keys  | Should -Contain "Test-FirstModuleFunction"
+        $module.ExportedFunctions.Keys  | Should -Not -Contain "Test-SecondModuleFunction"
+
+        $newModuleContent = (Get-Content $modData.Path | Out-String) + "`nfunction Test-SecondModuleFunction { Write-Output 'SECONDSTRING' }"
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: > Get-Content $modData.Path | Out-String

Why not use `$simpleModContent`? Or having `$newModuleContent` declared in `beforeall` like `$simpleModContent`?

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    It "Re-imports modules from file with new members when -Force is used with -PassThru" -Skip {
+        $module = Import-Module $modInfo -PassThru
+
+        $module.ExportedFunctions.Count | Should -Be 1
+        $module.ExportedFunctions.Keys  | Should -Contain "Test-FirstModuleFunction"
+        $module.ExportedFunctions.Keys  | Should -Not -Contain "Test-SecondModuleFunction"
+
+        $newModuleContent = (Get-Content $modData.Path | Out-String) + "`nfunction Test-SecondModuleFunction { Write-Output 'SECONDSTRING' }"
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Same comment here.

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,518 @@
...
+    }
+
+    It "Re-imports modules from file with new members when -Force is used with -AsCustomObject" -Skip {
+        $module = Import-Module $modInfo -PassThru
+
+        $module.ExportedFunctions.Count | Should -Be 1
+        $module.ExportedFunctions.Keys  | Should -Contain "Test-FirstModuleFunction"
+        $module.ExportedFunctions.Keys  | Should -Not -Contain "Test-SecondModuleFunction"
+
+        $newModuleContent = (Get-Content $modData.Path | Out-String) + "`nfunction Test-SecondModuleFunction { Write-Output 'SECONDSTRING' }"
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Ditto

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,518 @@
...
+Describe "Import-Module with nested modules" -Tag "Feature" {
+    BeforeAll {
+        SetupModules
+    }
+
+    AfterAll {
+        TearDownModules
+    }
+
+    It "Resolves submodule classes with 'using module'" {
```
Comment (daxian-dbw) [inline test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: I don't think `using module` is the right scenario for testing `import-module -force` doesn't refresh nested modules.
`using module` internally loads the module without `-Force`, so by definition forcing reloading the outer module doesn't refresh the ones pulled in by `using module`.

The right scenario should be nested modules or required modules declared in the module manifest I think.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]

**Comment by daxian-dbw** [comment]
This code should enable a developer scenario, where I have the following script in editor:
```powershell
using module F:\tmp\test.psm1

class Zoo { [Foo]$Param }
```
`F:\tmp\test.psm1` has the following content:
```powershell
class Foo { [string]$Name }
```
Then I add another class to `test.psm1` as follows:
```powershell
class Foo { [string]$Name }
class Bar { [string]$Path }
```
And then I modify `Zoo` in the editor to add a property `[Bar]$bar`:
```powershell
using module F:\tmp\test.psm1

class Zoo { [Foo]$Param; [Bar]$bar }
```
The `[Bar]` should be resolved in this scenario, which means the new content in `test.psm1` should be picked up by the parser.
Be noted, no module is loaded in this scenario.

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 (hunk 12 above)
Comment (rjmholt) [reply test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Well my intent here isn't to test `-Force` behaviour, just how `Import-Module` works in any scenarios with nested modules (seeing as how this seems to be a complicated area). The `using module` tests are just there to characterise the behaviour. We could move them to their own describe block though. @alx9r had a really nice breakdown of what tests he thought should be written in the old PR.

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 (hunk 7 above)
Comment (rjmholt) [reply test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: It does with `-Force` though

Comment (rjmholt) [reply test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: Is that the behaviour we are looking for?

File: test/powershell/engine/Module/ModuleCmdlet.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,518 @@
...
+'@
+
+            $modData = New-TestModule -Content $content
+
+            Import-Module $modData.Path -ArgumentList 'value1'
+            $rootVal = Get-PassedArgsRoot
+            $noRootVal = Get-PassedArgsNoRoot
+            $rootVal | Should -BeExactly $noRootVal
+
+            Import-Module $modData.Path -ArgumentList 'value2'
```
[the remark this answers is not in the document]
Comment (rjmholt) [reply test/powershell/engine/Module/ModuleCmdlet.Tests.ps1]: You're quite right that it doesn't.

I've duplicated the test to describe the normal and `-Force` cases (with a test to confirm the value of `$rootVal`).

**Comment by rjmholt** [comment]
@daxian-dbw I'm going to write another test in line with the example you give above

**Comment by rjmholt** [comment]
That test failure seems to be Linux-specific, and only occurs about 10% of the time.

**Comment by powercode** [comment]
I don't think this should be forgotten. Is a comment enough to keep it alive?

**Comment by rjmholt** [comment]
@powercode I promise I haven't forgotten about this PR... Just had to cut it short and haven't had a good reason to return to it. I'm about to get back to it - it looks like the failure is consistent on Linux, so I'm hoping to be able to get to the bottom of it,

**Comment by rjmholt** [comment]
Going to close this and continue working on it in an updated branch

**Pull request closed without merging**